source/g4collimation/G4Interface.cpp

```diff
@@ -154,34 +185,61 @@ extern "C" void g4_set_collimator_(char* name)
 	runManager->ReinitializeGeometry();
 }
 
-extern "C" void g4_collimate_(double* x, double* y, double* xp, double* yp, double* p)
+extern "C" void g4_add_particle_(double* x, double* y, double* xp, double* yp, double* e, int32_t* pdgid, int16_t* nzz, int16_t* naa, int16_t* nqq, double* mass)
 {
 //WARNING: at this stage in SixTrack the units have been converted to GeV, m, and rad!
 //The particle energy input is the TOTAL energy
+//mass (i.e. nucm) is already in MeV!
+
 	double x_in = (*x) * CLHEP::m;
 	double y_in = (*y) * CLHEP::m;
 
 //We want px and py, not the angle!
-	//double px_in = *p * tan(*xp) * CLHEP::GeV;
-	//double py_in = *p * tan(*yp) * CLHEP::GeV;
-	double px_in = (*p) * (*xp) * CLHEP::GeV;
-	double py_in = (*p) * (*yp) * CLHEP::GeV;
-	double p_in = (*p) * CLHEP::GeV;
+	double e_in = (*e) * CLHEP::GeV;
+	double p_in = sqrt((e_in*e_in) - (*mass * *mass));
+
+	double px_in = p_in * (*xp);
```

Comment:
This is probably not actually p_x -- here we have p_total times x'?

p_in^2 = (x'*p_z)^2 + (y'*p_z)^2 + p_z^2

solve for p_z

Must verify that x' is really angle!

Reply:
@rdemaria and @tpersson would be the best people to comment.

I think it is just p_x / p_0, with p_0 being the total momentum in whatever direction it is going.

Reply:
ok after checking with @rdemaria it actually is x' = p_x / p_total so it is NOT an angle, just a ratio, therefore it was correct before.

I've poked the code a bit and added some comments.